external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala

```diff
@@ -930,6 +930,33 @@ class AvroSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
     }
   }
 
+  test("support user provided non-nullable avro schema " +
+    "for nullable catalyst schema without any null record") {
```

Comment:
does parquet/orc have the same behavior? It seems better to forbid this at the beginning, otherwise we need to do null check at runtime, which may fail a long-running query middle away.

Reply:
cc @dongjoon-hyun @gengliangwang

Reply:
For `.schema()` option, we always enforce `nullable` by using `dataSchema.asNullable` in `FileTable`. For me, this is a special support for `.option("avroSchema", "")`.

Reply:
@cloud-fan I think this is fine. Otherwise, that is no way for users to write with a non-nullable schema.
But should we show warning messages for such case? So that users can be aware of the risk.

Reply:
Sounds good to have warning messages for the case. So can let users know they're actually writing from nullable catalyst schema into non-nullable avro schema.

Reply:
Adding warning message is good idea for me.

If we want to forbid this, we need to load the Avro with correct schema from the beginning. Currently, all the avro files will be loaded as `nullable == true` in catalyst schema even it's not nullable in the original avro schema. The round-trip of the loading and writing can result different schemas. This is the only workaround to write a non-nullable schema in avro easily to maintain the same schema.